docs/architecture/adr-026-protobuf-regencode.md

```diff
@@ -0,0 +1,223 @@
+# ADR 026: Protocol Buffer Regencode
+
+## Changelog
+
+- 2020-08-07: Initial Draft
+
+## Status
+
+Proposed
+
+## Context
+
+[Protobuf](https://developers.google.com/protocol-buffers/docs/proto3)
+seralization is not unique (i.e. there exist a practically unlimited number of
+valid binary representations for a protobuf document)<sup>1</sup>. For signature
+verification in Cosmos SDK, signer and verifier need to agree on the same
+serialization of a SignDoc as defined in
+[ADR-020](./adr-020-protobuf-transaction-encoding.md) without transmitting the
+serialization. This document describes a deterministic serialization scheme for
+a subset of protobuf documents, that covers this use case but can be reused in
+other cases as well.
+
+## Decision
+
+The following encoding scheme (called Regencode (verb), version 1) is proposed
+to be used by other ADRs.
+
+### Scope
+
+Regencode is a protobuf3 serializer. The output is a valid protobuf
+serialization, such that every protobuf parser can parse it.
+
+No maps are supported in version 1 due to the complexity of defining a
+derterministic serialization. This might change in future. Implementations must
+reject documents containing maps as invalid input.
+
+### Serialization rules
+
+The serialization is based on the
+[protobuf 3 encoding](https://developers.google.com/protocol-buffers/docs/encoding)
+with the following additions:
+
+1. Fields must be serialized in ascending order
+2. Extra fields or any extra data must not be added
+3. [Default values](https://developers.google.com/protocol-buffers/docs/proto3#default)
+   must be omitted
+
+While rule number 1. and 2. should be pretty straight forward and describe the
+default behaviour of all protobuf encoders the author is aware of, the 3rd rule
+is more interesting. Protobuf 3 has no notion of optional values and thus cannot
+differentiate between null/unset and empty fields. At serialization level
+however, it is possible to set the fields with an empty value or omitting them
+entirely. This is a significat difference to e.g. JSON where a property can be
+empty (`""`, `0`), `null` or undefined, leading to 3 different documents.
+
+Omitting empty fields is a valid option because the parser must assign the
+default value to fields missing in the serialization<sup>2</sup>. Requiring to
+serialize them would be equally valid. No good reason is known to the author for
+preferring one of those options over the other.
```

Comment:
```suggestion
Omitting empty fields is the logical choice for a canonical protobuf serialization because it allows for some amount of forward compatibility. That means that users of newer versions of a protobuf schema will produce the same serialization as users of older versions if they do not set newer fields. Effectively omitting newer fields causes the newer clients to fall back to an older version of the protocol. This would not be possible if empty fields were serialized by default.
```